norns/dev-setup.md

```diff
@@ -0,0 +1,146 @@
+nav_order: 3
+has_toc: false
+---
+
+# norns dev setup
+
+if using [maiden](../maiden) is plenty sufficient to create norns script, some experienced users may want to rely on development tools they are already familiar with.
+
+
+## remote access via command line
+
+### basics
+
+when accessing norns via the command line (using [SSH / USB serial access](../advanced-access)), scripts can directly be edited this way using locally-installed editors (such an `nano` or `vi`).
+
+this also allows interacting with [maiden from the command line](../maiden/#advanced-access)
+
+this video is a walkthrough of this process:
+
+<iframe width="560" height="315" src="https://www.youtube.com/embed/S-jhIT4mCcM" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+
+
+### sshfs - mount source files locally
+
+`sshfs` allow to locally mount norns filesystem onto a development machine as if it were local. this allows you to use any prefered editor / tools installed locally.
+
+```shell
+mkdir /home/Documents/norns_mount
+cd /home/Documents/norns_mount
+shfs we@norns:/home/we/dust dust
+```
+
+please note that many editors provide an equivalent functionality.
+
+
+### maiden-run - automatic reload
+
+[`maiden-run`](https://github.com/ngwese/maiden-run) allows to automatically reload  a script when saving its source code.
+
+
+## editor integration
+
+### atom
+
+this [comment](https://github.com/monome/norns/issues/1067#issuecomment-611732427) describes a working setup for using `atom` on a computer to access and code on a remote norns.
+
+
+### emacs
+
+to enable lua and supercollider support:
+
+```elisp
+;; lua
+
+(use-package lua-mode
+  :init
+  (setq lua-indent-level 2))
+
+;; sclang
+
+(defvar my-supercollider-scel-path "~/.local/share/SuperCollider/Emacs/scel")
+
+(let ((sclang-dir (concat my-supercollider-scel-path "/el")))
+  (when (file-directory-p sclang-dir)
+    (normal-top-level-add-to-load-path (list sclang-dir))))
+
+(when (and (locate-library "sclang")
+           (locate-library "sclang-vars"))
+  (use-package sclang
+    :ensure nil
+    :demand))
+```
+
+Emacs can access norns filesystem as if local using [TRAMP](https://www.gnu.org/software/tramp/).
+
+```shell
+	M-x find-file
+  /ssh:we@norns.local:/home/we/dust/
+```
+
+to prevent having to re-type the password each time, put this line in your `~/.authinfo`:
+
+```conf
+machine norns.local login we port ssh password sleep
+```
+
+for spawning a maiden repl, do it using either command:
+ - `tramp-term` (built-in `term` + [tramp-term](https://github.com/randymorris/tramp-term.el))
+ - `vterm-toggle-cd` ([vterm](https://github.com/akermu/emacs-libvterm) + [vterm-toggle](https://github.com/jixiuf/vterm-toggle))
+
+
+### VSCode
+
+norns can be accessed as if locally using [Remote - SSH extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-ssh).
```

Comment:
using the remote vscode backend over ssh consumes a considerable amount of disk space, memory, and cpu given that it downloads and runs a full nodejs environment. the resource usage is high enough that i would expect it to interfere with the norns stack - i've personally avoided it for this very reason. i would recommend users use smb + vscode running locally instead.

Reply:
oh, so it needs a nodejs instance running on norns?!

i would have thought it'd be more transparent (relying only on SSH like most other editors).

so we either:
- remove this recommendation
- keep it but w/ a big disclaimer about this resource consumption

Reply:
i would lean towards removing it as a recommendation.

the remote ssh backend basically installs a headless copy of vscode and nodejs on the remote device. when using the various "remote" extensions the desktop editor is essentially a client and _everything_ runs on the remote machine. this is great in cloud environments but a really bad fit for resource constrained embedded devices like norns.

Reply:
"downloads and runs a full nodejs environment." 😱 😱 😱 😱

Reply:
ok, removed it.

i've nonethless kept the `norns REPL` extension which is lightweight (no need to install anything, does everything through existing websocket connexion capability).